File: services/scala/src/main/scala/tests/RegexRedux.scala
package tests

import scala.util.matching.Regex

object RegexRedux {
  // Define regex patterns (case-insensitive)
  private val regExps = List(
    """(?i)agggtaaa|tttaccct""".r.unanchored,
    """(?i)[cgt]gggtaaa|tttaccc[acg]""".r.unanchored,
    """(?i)a[act]ggtaaa|tttacc[agt]t""".r.unanchored,
    """(?i)ag[act]gtaaa|tttac[agt]ct""".r.unanchored,
    """(?i)agg[act]taaa|ttta[agt]cct""".r.unanchored,
    """(?i)aggg[acg]aaa|ttt[cgt]ccct""".r.unanchored,
    """(?i)agggt[cgt]aa|tt[acg]accct""".r.unanchored,
    """(?i)agggta[cgt]a|t[acg]taccct""".r.unanchored,
    """(?i)agggtaa[cgt]|[acg]ttaccct""".r.unanchored
  )

  private val patternStrings = List(
    "agggtaaa|tttaccct",
    "[cgt]gggtaaa|tttaccc[acg]",
    "a[act]ggtaaa|tttacc[agt]t",
    "ag[act]gtaaa|tttac[agt]ct",
    "agg[act]taaa|ttta[agt]cct",
    "aggg[acg]aaa|ttt[cgt]ccct",
    "agggt[cgt]aa|tt[acg]accct",
    "agggta[cgt]a|t[acg]taccct",
    "agggtaa[cgt]|[acg]ttaccct"
  )

  def regexRedux(): String = {
    // Use a sample DNA sequence for testing since we can't read from stdin in a web context
    val sampleData =
      """
    >ONE Homo sapiens aluGGCCGGGCGCGGTGGCTCACGCCTGTAA
    AGGCGGGCGGATCACCTGAGGTCAGGAGTTCGAGACCAGCCTGGCCAAC
    ATGGTGAAACCCCGTCTCTACTAAAAATACAAAAATTAGCCGGGCGTGG
    TGGCGCATGCCTGTAATCCCAGCTACTCGGGAGGCTGAGGCAGGAGAAT
    CGCTTGAACCCGGGAGGCGGAGGTTGCAGTGAGCCGAGATCGCGCCACT
    GCACTCCAGCCTGGGCGACAGAGCGAGACTCCGTCTCAAAAAGGCCGGG
    CGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCCGAGGCGGG
    CGGATCACCTGAGGTCAGGAGTTCGAGACCAGCCTGGCCAACATGGTGA
    AACCCCGTCTCTACTAAAAATACAAAAATTAGCCGGGCGTGGTGGCGCA
    TGCCTGTAATCCCAGCTACTCGGGAGGCTGAGGCAGGAGAATCGCTTGA
    ACCCGGGAGGCGGAGGTTGCAGTGAGCCGAGATCGCGCCACTGCACTCC
    AGCCTGGGCGACAGAGCGAGACTCCGTCTCAAAAAGGCCGGGCGCGGTG
    GCTCACGCCTGTAATCCCAGCACTTTGGGAGGCCGAGGCGGGCGGATCA
    CCTGAGGTCAGGAGTTCGAGACCAGCCTGGCCAACATGGTGAAACCCCG
    TCTCTACTAAAAATACAAAAATTAGCCGGGCGTGGTGGCGCATGCCTGT
    AATCCCAGCTACTCGGGAGGCTGAGGCAGGAGAATCGCTTGAACCCGGG
    AGGCGGAGGTTGCAGTGAGCCGAGATCGCGCCACTGCACTCCAGCCTGG
    GCGACAGAGCGAGACTCCGTCTCAAAAA
  """

    var data = sampleData
    val initialLen = data.length

    // Remove lines starting with > and all newlines
    data = """(?m)^>.*\n|\n""".r.replaceAllIn(data, "")
    val cleanedLen = data.length

    // Count matches for each pattern (case-insensitive)
    val results = regExps.zip(patternStrings).map { case (re, patternStr) =>
      val matches = re.findAllIn(data).toList.length
      s"$patternStr $matches"
    }

    // Apply replacement chain
    val endLen = data
      .replaceAll("(?i)tHa[Nt]", "<4>")
      .replaceAll("(?i)aND|caN|Ha[DS]|WaS", "<3>")
      .replaceAll("(?i)a[NSt]|BY", "<2>")
      .replaceAll("<[^>]*>", "|")
      .replaceAll("\\|[^|][^|]*\\|", "-")
      .length

    val response = List(
      s"Initial length: $initialLen",
      s"Cleaned length: $cleanedLen",
      s"Final length: $endLen",
      "",
      "Pattern matches:"
    ) ++ results

    response.mkString("\n")
  }
}

File: services/python/src/tests/regex_redux.py
import re

def regex_redux():
    """Perform regex pattern matching and replacements on DNA sequence data"""
    reg_exps = [
        re.compile(r'agggtaaa|tttaccct', re.IGNORECASE),
        re.compile(r'[cgt]gggtaaa|tttaccc[acg]', re.IGNORECASE),
        re.compile(r'a[act]ggtaaa|tttacc[agt]t', re.IGNORECASE),
        re.compile(r'ag[act]gtaaa|tttac[agt]ct', re.IGNORECASE),
        re.compile(r'agg[act]taaa|ttta[agt]cct', re.IGNORECASE),
        re.compile(r'aggg[acg]aaa|ttt[cgt]ccct', re.IGNORECASE),
        re.compile(r'agggt[cgt]aa|tt[acg]accct', re.IGNORECASE),
        re.compile(r'agggta[cgt]a|t[acg]taccct', re.IGNORECASE),
        re.compile(r'agggtaa[cgt]|[acg]ttaccct', re.IGNORECASE)
    ]

    # Use a sample DNA sequence for testing since we can't read from stdin in a web context
    sample_data = """
    >ONE Homo sapiens aluGGCCGGGCGCGGTGGCTCACGCCTGTAA
    AGGCGGGCGGATCACCTGAGGTCAGGAGTTCGAGACCAGCCTGGCCAAC
    ATGGTGAAACCCCGTCTCTACTAAAAATACAAAAATTAGCCGGGCGTGG
    TGGCGCATGCCTGTAATCCCAGCTACTCGGGAGGCTGAGGCAGGAGAAT
    CGCTTGAACCCGGGAGGCGGAGGTTGCAGTGAGCCGAGATCGCGCCACT
    GCACTCCAGCCTGGGCGACAGAGCGAGACTCCGTCTCAAAAAGGCCGGG
    CGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCCGAGGCGGG
    CGGATCACCTGAGGTCAGGAGTTCGAGACCAGCCTGGCCAACATGGTGA
    AACCCCGTCTCTACTAAAAATACAAAAATTAGCCGGGCGTGGTGGCGCA
    TGCCTGTAATCCCAGCTACTCGGGAGGCTGAGGCAGGAGAATCGCTTGA
    ACCCGGGAGGCGGAGGTTGCAGTGAGCCGAGATCGCGCCACTGCACTCC
    AGCCTGGGCGACAGAGCGAGACTCCGTCTCAAAAAGGCCGGGCGCGGTG
    GCTCACGCCTGTAATCCCAGCACTTTGGGAGGCCGAGGCGGGCGGATCA
    CCTGAGGTCAGGAGTTCGAGACCAGCCTGGCCAACATGGTGAAACCCCG
    TCTCTACTAAAAATACAAAAATTAGCCGGGCGTGGTGGCGCATGCCTGT
    AATCCCAGCTACTCGGGAGGCTGAGGCAGGAGAATCGCTTGAACCCGGG
    AGGCGGAGGTTGCAGTGAGCCGAGATCGCGCCACTGCACTCCAGCCTGG
    GCGACAGAGCGAGACTCCGTCTCAAAAA
    """

    data = sample_data
    initial_len = len(data)

    data = re.sub(r'^>.*\n|\n', '', data, flags=re.MULTILINE)
    cleaned_len = len(data)

    results = []
    for reg_exp in reg_exps:
        matches = reg_exp.findall(data)
        pattern_str = reg_exp.pattern
        results.append(f'{pattern_str} {len(matches)}')

    data = re.sub(r'tHa[Nt]', '<4>', data, flags=re.IGNORECASE)
    data = re.sub(r'aND|caN|Ha[DS]|WaS', '<3>', data, flags=re.IGNORECASE)
    data = re.sub(r'a[NSt]|BY', '<2>', data, flags=re.IGNORECASE)
    data = re.sub(r'<[^>]*>', '|', data)
    data = re.sub(r'\|[^|][^|]*\|', '-', data)
    end_len = len(data)

    return '\n'.join([
        f'Initial length: {initial_len}',
        f'Cleaned length: {cleaned_len}',
        f'Final length: {end_len}',
        '',
        'Pattern matches:',
        *results
    ])

